official-antistasi-community/A3-Antistasi
# Convoy mission rework

## What type of PR is this.
1. [X] Bug
2. [X] Change
3. [ ] Enhancement

### What have you changed and why?
General rebalance and fixup to make convoy missions more plausible to complete:
- Target selection uses much closer targets so you don't have to drive 8km to reach the convoy.
- Source selection prefers medium distances to allow more intercept options.
- Convoy units spawn immediately but have a 5-20 minute travel delay.
- Convoy travel script for vehicles so that they actually follow each other in convoy.
- Convoy vehicle spawning now much more reliable.
- allowCrewInImmobile used on armoured vehicles for consistency.
- Ammo convoys now actually contain loot-crate equipment.
- Fixed typo in reinforcement convoys.
- Escort vehicles RTB at completion.
- groupDespawner now removes spawner flag.

The spawning code includes a lot of generalised pathfinding support code:
- Find suitable navpoint for marker.
- Road connecting point/intercept.
- Road-to-road A-star.
- Find spawn positions along route.

This ended up with a lot more code and complexity than the original intention, as all reasonably simple spawning mechanisms had massive flaws. On the plus side, most of the pathfinding stuff should be generally useful, at least as examples.

Remaining issues:
- Disabled the convoy icons (inmuneconvoy) as the function does some potentially-problematic stuff that isn't icon-specific. Might add it back in shortly.
- The nearest navgrid point to a marker or spawnpoint can be quite far from the marker (common with factories and resources placed on a straight road), or in the case of spawnpoints, deep in a nest of roads that we'd rather not touch (see the middle airport on Altis). The first one could be fixed fairly cheaply on load, but I'm not sure how to handle the second.
- Arma AIs drive extremely poorly in many situations. Having the vehicles point in the right direction and avoid overtaking helps a lot, but other cases are still bad. Navgrid tweaks may help a bit, but not much.
- Convoys are probably too strong in terms of enemy firepower versus rewards. Depends how many vehicles get stuck.

### Please specify which Issue this PR Resolves.
closes #1745
closes #1568
closes #1492

### Please verify the following and ensure all checks are completed.
1. [X] Have you loaded the mission in LAN host?
2. [X] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [X] No
2. [ ] Yes (Please provide further detail below.)

### How can the changes be tested?
You can request convoy missions from Petros to check the source/target selection results. 

Force-creating specific convoys. Run as server:
`[destMarker, sourceMarker, convoyType] spawn A3A_fnc_convoy;`
Convoy type can be left nil or empty string to autoselect, or choose a type from early in the function.
